File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.res=0
        def helper(node):
            if(node==None):
                return 0
            a=helper(node.left)
            b=helper(node.right)

            self.res=max(self.res, a+b)

            return max(a, b)+1
        
        p=helper(root.left) + helper(root.right)

        return max(self.res,p)

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let res=0
    function helper(node){
        if(node===null) return 0
        let a=helper(node.left)
        let b=helper(node.right)
        
        res=Math.max(res, a+b)
        // console.log(node.val, a, b, res)
        return Math.max(a, b)+1
    }
    let p=helper(root.left) + helper(root.right)

    return Math.max(res,p)
};